Firestore/core/src/firebase/firestore/local/serializer.cc

```diff
@@ -0,0 +1,182 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/local/serializer.h"
+
+#include <cstdlib>
+#include <utility>
+
+#include "Firestore/Protos/nanopb/firestore/local/maybe_document.nanopb.h"
+#include "Firestore/Protos/nanopb/google/firestore/v1beta1/document.nanopb.h"
+#include "Firestore/core/src/firebase/firestore/model/field_value.h"
+#include "Firestore/core/src/firebase/firestore/model/no_document.h"
+#include "Firestore/core/src/firebase/firestore/nanopb/tag.h"
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+using firebase::firestore::model::ObjectValue;
+using firebase::firestore::nanopb::Reader;
+using firebase::firestore::nanopb::Tag;
+using firebase::firestore::nanopb::Writer;
+using firebase::firestore::util::Status;
+
+Status Serializer::EncodeMaybeDocument(const model::MaybeDocument& document,
+                                       std::vector<uint8_t>* out_bytes) const {
+  Writer writer = Writer::Wrap(out_bytes);
+  EncodeMaybeDocument(&writer, document);
+  return writer.status();
+}
+
+void Serializer::EncodeMaybeDocument(
+    Writer* writer, const model::MaybeDocument& maybe_doc) const {
+  switch (maybe_doc.type()) {
+    case model::MaybeDocument::Type::Document:
+      writer->WriteTag(
+          {PB_WT_STRING, firestore_client_MaybeDocument_document_tag});
+      writer->WriteNestedMessage([&](Writer* writer) {
+        EncodeDocument(writer, static_cast<const model::Document&>(maybe_doc));
+      });
+      return;
+
+    case model::MaybeDocument::Type::NoDocument:
+      // TODO(rsgowman)
+      abort();
+
+    case model::MaybeDocument::Type::Unknown:
+      // TODO(rsgowman)
+      abort();
+  }
+
+  UNREACHABLE();
+}
+
+std::unique_ptr<model::MaybeDocument> Serializer::DecodeMaybeDocument(
+    Reader* reader) const {
+  if (!reader->status().ok()) return nullptr;
+
+  // Initialize MaybeDocument fields to their default values. (Due to the
+  // 'oneof' in MaybeDocument, only one of 'no_document' or 'document' should
+  // ever be set.)
+  std::unique_ptr<model::NoDocument> no_document;
+  std::unique_ptr<model::Document> document;
+
+  while (reader->bytes_left()) {
+    Tag tag = reader->ReadTag();
+    if (!reader->status().ok()) return nullptr;
+
+    // Ensure the tag matches the wire type
+    switch (tag.field_number) {
+      case firestore_client_MaybeDocument_document_tag:
+      case firestore_client_MaybeDocument_no_document_tag:
+        if (tag.wire_type != PB_WT_STRING) {
+          reader->set_status(
```

Comment:
Is there any reason not to just return immediately after setting status?

I worry that by not returning early we leave ourselves open to the possibility that a future refactoring will break the return that's currently below.

Additionally it makes the error handling (slightly) harder to reason about since the return value and the status generation aren't paired.

These aren't strong arguments, but it really isn't any extra code to bind these more closely.

Reply:
> Is there any reason not to just return immediately after setting status?

Not that I can recall. I've moved it in here (and will do so throughout the rest of the serializers as I touch them).